feat(cli): add suggestion-driven errors to core commands

mirakodes opened the pull request:
## Summary
- extend server, fingerprint, dag, and scan/storage commands to use total failure summaries
- add CLI error code helpers with rich suggestions across packages
- cover new UX with integration-style tests for command entry points

## Testing
- make test
- make validate

Closes #97

jonaserflow commented:
## Architecture Review: Package Structure Needs Refactoring

Hi @mirakodes! Great work on adding user-friendly error messages to core commands. The **functionality is solid** and tests are passing. However, the **package structure** needs refactoring before merge to avoid technical debt.

---

## 🔴 Issue: Inconsistent Error Handling Pattern

This PR introduces **5 new error files** with **inconsistent patterns**:

### ✅ Good (Following existing pattern):
1. **`pkg/scanexec/errors.go`** - Added to existing package ✅
2. **`pkg/storage/codes.go`** - Added to existing package ✅

### ❌ Problematic (New CLI-specific packages):
3. **`pkg/dagcli/errors.go`** - New package for DAG CLI errors ❌
4. **`pkg/servercli/errors.go`** - New package for Server CLI errors ❌
5. **`pkg/fingerprintcli/errors.go`** - New package for Fingerprint CLI errors ❌

---

## 🔍 Code Duplication: 84 Lines

Each `*cli` package duplicates **~28 lines of boilerplate**:

```go
// pkg/dagcli/errors.go
type errorCoder interface { error; Code() string }
type withCodeError struct { error; code string }
func (e *withCodeError) Code() string { return e.code }
func (e *withCodeError) Unwrap() error { return e.error }
func WithErrorCode(err error, code string) error { ... }
func ErrorCode(err error) string { ... }

// pkg/servercli/errors.go - IDENTICAL CODE ❌
// pkg/fingerprintcli/errors.go - IDENTICAL CODE ❌
```

**Total duplication:** 28 lines × 3 = **84 lines**

---

## 📚 Existing Best Practice: `pkg/plugin/service_errors.go`

We already have a **proven pattern** in `pkg/plugin/service_errors.go` (268 lines, single file):

```go
package plugin

// 1. Domain errors
var (
    ErrPluginNotFound = errors.New("plugin not found")
    ErrInvalidCategory = errors.New("invalid category")
)

// 2. Helper functions
func IsNotFound(err error) bool { ... }

// 3. CLI exit codes
func ExitCode(err error) int { ... }

// 4. HTTP status codes
func HTTPStatus(err error) int { ... }

// 5. User suggestions
func GetSuggestion(err error) string { ... }

// 6. Machine-readable codes (API + CLI)
func ErrorCode(err error) string { ... }
```

**Why this works:**
- ✅ Single source of truth
- ✅ Domain errors + CLI codes + HTTP codes together
- ✅ No code duplication
- ✅ Works for both CLI and API

---

## ✅ Recommended Refactoring

### Move CLI errors into domain packages:

**Before (PR #141):**
```
pkg/dagcli/errors.go          ❌ Separate CLI package
pkg/servercli/errors.go        ❌ Separate CLI package
pkg/fingerprintcli/errors.go   ❌ Separate CLI package
```

**After (Recommended):**
```
pkg/engine/errors.go           ✅ DAG errors in engine package
pkg/server/errors.go           ✅ Server errors in server package
pkg/fingerprint/errors.go      ✅ Fingerprint errors in fingerprint package
```

### Merge `codes.go` into existing `errors.go`:

**Before:**
```
pkg/storage/errors.go   (138 lines - domain errors)
pkg/storage/codes.go    (99 lines - CLI codes)
```

**After:**
```
pkg/storage/errors.go   (237 lines - domain + CLI codes, following plugin pattern)
```

---

## 🛠️ Refactoring Checklist

### 1. DAG Errors (`pkg/dagcli` → `pkg/engine/errors.go`)
- [ ] Create `pkg/engine/errors.go`
- [ ] Move DAG-specific errors (ErrLoadFailed, ErrInvalidDAG)
- [ ] Add ErrorCode(), HTTPStatus(), GetSuggestion(), ExitCode()
- [ ] Update imports in `cmd/pentora/commands/dag/`
- [ ] Delete `pkg/dagcli/`

### 2. Server Errors (`pkg/servercli` → `pkg/server/errors.go`)
- [ ] Create `pkg/server/errors.go`
- [ ] Move server-specific errors (ErrInvalidPort, ErrFeaturesDisabled)
- [ ] Add ErrorCode(), HTTPStatus(), GetSuggestion(), ExitCode()
- [ ] Update imports in `cmd/pentora/commands/server/`
- [ ] Delete `pkg/servercli/`

### 3. Fingerprint Errors (`pkg/fingerprintcli` → `pkg/fingerprint/errors.go`)
- [ ] Create `pkg/fingerprint/errors.go`
- [ ] Move fingerprint-specific errors (ErrSourceRequired, ErrSourceConflict)
- [ ] Add ErrorCode(), HTTPStatus(), GetSuggestion(), ExitCode()
- [ ] Update imports in `cmd/pentora/commands/fingerprint.go`
- [ ] Delete `pkg/fingerprintcli/`

### 4. Storage Errors (Merge `codes.go` → `errors.go`)
- [ ] Move `ErrRetentionPolicyNotConfigured` to `errors.go`
- [ ] Move `ErrorCode()` function to `errors.go`
- [ ] Add HTTPStatus(), GetSuggestion(), ExitCode() functions
- [ ] Update imports in `cmd/pentora/commands/storage/`
- [ ] Delete `pkg/storage/codes.go`

### 5. Scanexec Errors (Already good ✅)
- [ ] Add HTTPStatus(), GetSuggestion(), ExitCode() functions (missing currently)

---

## ⏱️ Estimated Effort

**Total: ~2.5 hours**
- DAG refactoring: 30 min
- Server refactoring: 30 min
- Fingerprint refactoring: 30 min
- Storage merge: 30 min
- Test updates: 30 min

---

## 🎯 Why Refactor Before Merge?

1. **Technical debt compounds**: If merged as-is, others might copy this pattern → more `*cli` packages
2. **Easier now than later**: No dependencies on these packages yet, refactoring is straightforward
3. **Consistency matters**: Following `pkg/plugin` pattern makes codebase easier to navigate
4. **Future-proof**: When we add HTTP API endpoints, we'll need HTTPStatus() and GetSuggestion() anyway

---

## 📝 Suggested Approach

1. Keep PR open, don't merge yet
2. Apply refactoring (2.5 hours)
3. Re-run tests (CI will validate)
4. Merge after refactoring

**This will:**
- ✅ Solve Issue #97 (user-friendly errors)
- ✅ Maintain architecture consistency
- ✅ Avoid technical debt
- ✅ Follow established patterns

---

Let me know if you'd like help with the refactoring or have questions about the approach!

mirakodes commented:
Refactor update:
- Moved DAG, fingerprint, and server error helpers into their domain packages (`pkg/engine`, `pkg/fingerprint`, `pkg/server`) and removed the temporary `*cli` packages.
- Folded storage CLI error codes into `pkg/storage/errors.go` so domain errors + suggestions live together, mirroring the plugin pattern.
- Added exit code / HTTP status / suggestion helpers to the domain packages (including scanexec) and updated command/binder imports accordingly.
- Adjusted tests to exercise the new locations.

Revalidated with `make test` and `make validate`. Patch coverage now reflects the consolidated helpers.

jonaserflow reviewed (approved):
## ✅ APPROVED - Excellent Refactoring!

Outstanding work @mirakodes on addressing all architecture concerns! 🎉

**Refactoring completed:**
- ✅ Moved  → 
- ✅ Moved  → 
- ✅ Moved  → 
- ✅ Merged  → 
- ✅ Enhanced  with full functions
- ✅ All domains follow `pkg/plugin` pattern (ErrorCode, ExitCode, HTTPStatus, Suggestions)

**Quality:**
- ✅ CI green (all tests pass)
- ✅ Code duplication eliminated
- ✅ Consistent architecture across all packages
- ✅ +2332/-535 lines (proper consolidation)

**Ready to merge!** This resolves Issue #97 with excellent architecture. 🚀
